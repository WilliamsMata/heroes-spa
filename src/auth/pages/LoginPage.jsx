import { useContext } from "react";
import { useNavigate } from "react-router-dom";

import { AuthContext } from "../context/AuthContext";

export const LoginPage = () => {
  const { login } = useContext(AuthContext);
  const navigate = useNavigate();

  const onLogin = () => {
    const lastPath = localStorage.getItem("lastPath") || "/";
    login("Williams Mata");

    navigate(lastPath, {
      replace: true,
    });
  };

  return (
    <>
      <div className="container-lg">
        <h1 className="mt-2">Login</h1>
        <hr />
        <button className="btn btn-primary" onClick={onLogin}>
          Login
        </button>
      </div>
    </>
  );
};
