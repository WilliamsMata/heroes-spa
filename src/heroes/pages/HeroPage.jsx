import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroById } from "../helpers";

export const HeroPage = () => {
  const { id } = useParams();
  const hero = useMemo(() => getHeroById(id), [id]);

  const navigate = useNavigate();
  const onNavigateBack = () => navigate(-1);

  if (!hero) {
    return <Navigate to={"/marvel"} />;
  }

  return (
    <div
      className="row mt-2"
      style={{ marginInline: "auto", maxWidth: "62.5rem" }}
    >
      <div className="col-12 col-sm-4 animate__animated animate__fadeInLeft">
        <img
          src={`/images/heroes/${id}.jpg`}
          alt={hero.superhero}
          className="img-thumbnail"
        />
      </div>

      <div className="col-12 col-sm-8">
        <h3>{hero.superhero}</h3>

        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego:</b> {hero.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher:</b> {hero.publisher}
          </li>
          <li className="list-group-item">
            <b>First Appearance:</b> {hero.first_appearance}
          </li>
        </ul>

        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>

        <button
          onClick={onNavigateBack}
          className="btn btn-outline-primary mb-2 "
        >
          Back
        </button>
      </div>
    </div>
  );
};
