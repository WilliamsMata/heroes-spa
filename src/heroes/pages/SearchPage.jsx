import { useNavigate, useLocation } from "react-router-dom";
import queryString from "query-string";

import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../components";
import { getHeroesByName } from "../helpers";

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);
  const heroes = getHeroesByName(q);

  const showSearch = q.length === 0;
  const showError = q.length > 0 && heroes.length === 0;

  const { searchText, onInputChange, onResetForm } = useForm({
    searchText: "",
  });

  const onSearchSubmit = (event) => {
    event.preventDefault();
    navigate(`?q=${searchText}`);
    onResetForm();
  };

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-12 col-sm-5 animate__animated animate__fadeIn">
          <form onSubmit={onSearchSubmit}>
            <input
              type="text"
              placeholder="Search Hero"
              name="searchText"
              autoComplete="off"
              className="form-control "
              value={searchText}
              onChange={onInputChange}
            />

            <button className="btn btn-outline-primary mt-2">Search</button>
          </form>
        </div>

        <div className="col-12 col-sm-7 mt-2 mt-sm-0">
          <div
            className="alert alert-primary animate__animated animate__fadeIn"
            style={{ display: showSearch ? "" : "none" }}
          >
            Search a Hero
          </div>

          <div
            className="alert alert-danger animate__animated animate__fadeIn"
            style={{ display: showError ? "" : "none" }}
          >
            There's no hero with <b>{q}</b>
          </div>

          {heroes.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
          {/* <HeroCard /> */}
        </div>
      </div>
    </>
  );
};
