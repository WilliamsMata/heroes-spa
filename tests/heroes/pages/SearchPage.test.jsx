import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("Pruebas en <SearchPage />", () => {
  beforeEach(() => jest.clearAllMocks());

  test("Debe de mostrarse correctamente con valores por defecto", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test("Debe de mostrar a Batman", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />
      </MemoryRouter>
    );
    const img = screen.getByRole("img");
    const searchHeroText = screen.getByLabelText("search-hero-text");
    const noHeroWithText = screen.getByLabelText("no-hero-text");

    expect(img.src).toContain("images/heroes/dc-batman.jpg");
    expect(searchHeroText.style.display).toBe("none");
    expect(noHeroWithText.style.display).toBe("none");
  });

  test("Debe de mostrar un error si no se encuentra el hero (batman123)", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman123"]}>
        <SearchPage />
      </MemoryRouter>
    );
    const searchHeroText = screen.getByLabelText("search-hero-text");
    const noHeroWithText = screen.getByLabelText("no-hero-text");

    expect(searchHeroText.style.display).toBe("none");
    expect(noHeroWithText.style.display).not.toBe("none");
  });

  test("Debe de llamar el navigate a la pantalla nueva", () => {
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const input = screen.getByRole("textbox");
    const form = screen.getByLabelText("form");

    fireEvent.input(input, { target: { value: "batman" } });
    fireEvent.submit(form);

    expect(mockedUseNavigate).toHaveBeenCalledWith("?q=batman");
  });
});
