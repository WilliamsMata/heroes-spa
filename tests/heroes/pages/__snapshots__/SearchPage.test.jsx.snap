// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-12 col-sm-5 animate__animated animate__fadeIn"
    >
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control "
          name="searchText"
          placeholder="Search Hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-12 col-sm-7 mt-2 mt-sm-0"
    >
      <div
        aria-label="search-hero-text"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a Hero
      </div>
      <div
        aria-label="no-hero-text"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        There's no hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
